app/src/main/java/org/ea/sqrl/activites/base/LoginBaseActivity.java

```diff
@@ -240,47 +201,58 @@ public void afterTextChanged(Editable s) {
             long currentId = sharedPref.getLong(CURRENT_ID, 0);
 
             if(currentId != 0) {
-                hideLoginPopup();
-                showProgressPopup();
-                closeKeyboard();
-
-                new Thread(() -> {
-                    boolean decryptionOk = storage.decryptIdentityKey(txtLoginPassword.getText().toString(), entropyHarvester, false);
-                    if(!decryptionOk) {
-                        showErrorMessage(R.string.decrypt_identity_fail);
-                        handler.post(() -> {
-                            txtLoginPassword.setText("");
-                            hideProgressPopup();
-                        });
-                        storage.clearQuickPass();
-                        storage.clear();
-                        return;
-                    }
-                    showClearNotification();
-
-                    handler.post(() -> txtLoginPassword.setText(""));
-
-                    communicationFlowHandler.addAction(CommunicationFlowHandler.Action.QUERY_WITHOUT_SUK_QRCODE);
-                    communicationFlowHandler.addAction(CommunicationFlowHandler.Action.LOGIN);
-
-                    communicationFlowHandler.setDoneAction(() -> {
-                        storage.clear();
-                        handler.post(() -> {
-                            hideProgressPopup();
-                            closeActivity();
-                        });
-                    });
-
-                    communicationFlowHandler.setErrorAction(() -> {
-                        storage.clear();
-                        storage.clearQuickPass();
-                        handler.post(() -> hideProgressPopup());
-                    });
-
-                    communicationFlowHandler.handleNextAction();
-
-                }).start();
+                doLogin(storage, txtLoginPassword, false, false, null, this);
             }
         });
     }
+
+    public void doLogin(SQRLStorage storage, EditText txtLoginPassword, boolean usedQuickpass, boolean usedCps, Activity activityToFinish, Context context) {
+        if (!usedCps) hideLoginPopup();
+        showProgressPopup();
+        closeKeyboard();
+
+        new Thread(() -> {
+            boolean decryptionOk = storage.decryptIdentityKey(txtLoginPassword.getText().toString(), entropyHarvester, usedQuickpass);
+            if(!decryptionOk) {
+                showErrorMessage(R.string.decrypt_identity_fail);
+                handler.post(() -> {
+                    txtLoginPassword.setHint(R.string.login_identity_password);
+                    txtLoginPassword.setText("");
+                    hideProgressPopup();
+                    showLoginPopup();
+                });
+                storage.clear();
+                storage.clearQuickPass();
+                return;
+            }
+            if(!usedCps) showClearNotification();
```

Comment:
I made a note in the code where `showClearNotification()` wasn't being called.

This would change the behavior compared to existing code, but I can see how it needs to be removed so it will always be called.  I'll work on it.